### PR TITLE
Feature: Tag templates

### PR DESCRIPTION
**Pull request by cortadocodes**
### New Features
- [x] Make `Datafile`s and `Dataset`s labelable
- [x] Use new version of tags in all `Taggable`s
- [x] Replace string tags in a `TagSet` with key-value pairs in a `TagDict`
- [x] Add new `Taggable` mixin for providing the new tags interface
- [x] Add `FilterDict`, allowing filtering of key-value pairs by their values
- [x] Allow nested attribute/dictionary filtering in the filter containers `FilterSet`, `FilterList` and `FilterDict`
- [x] Allow any number of filters to be specified when filtering in filter containers
- [x] Allow ignoring of filterables missing the filtered-for attribute in a filter container instead of raising an error
- [x] Add filter container `one` method? 
- [x] Allow ordering by nested attributes in all `FilterContainer`s

### Breaking changes
- [x] Use new format for manifests' datasets in `twine.json` files 
- [x] Convert old `Taggable` mixin to `Labelable` mixin
- [x] Convert old `Tag` class to `Label` class
- [x] Convert `TagSet` to `LabelSet` 
- [x] Use key-value pairs for filter names and values when filtering `Filterable`s
- [x] Stop logging in `Serialisable`
- [x] Always exclude `logger` field in `Serialisable`
- [x] Simplify tag name pattern to `^[a-z0-9][a-z0-9_]*(?<!_)$`
- [x] Simplify label pattern to `^[a-z0-9][a-z0-9-]*(?<!-)$`
- [x] Store tags as key-value pairs in GCS custom metadata 
- [x] Unbase `TagDict` and `LabelSet` from filter containers
- [x] JSON-encode cloud storage custom metadata again
- [x] Store tags in `tags` field of cloud metadata again
 
### Minor improvements
- [x] Remove `filters` field from manifest strand in twines
- [x] Allow tags to be added via `kwargs` in `Taggable.add_tags`
- [x] Remove unused `_FILTERSET_ATTRIBUTE` class variables
- [x] Base `Label` on `str` 
- [x] Support non-English characters in case-insensitive filtering 
- [x] Add `octue-sdk-python` version to datafile metadata
- [x] Base filter containers on new `FilterContainer` abstract class
- [x] Move `filter` and `order` methods into `FilterContainer`
- [x] Use `OctueJSONDecoder` in `Serialisable` and `GoogleCloudStorageClient`
- [x] Add de/serialisation of `datetime` objects to de/encoders
- [x] Clarify name of some `GoogleCloudStorageClient` methods
- [x] Add `set` and `UserString` encoding to `OctueJSONEncoder` 
- [x] Use `OctueJSONDecoder` 
- [x] Add `set` and `datetime` decoding to `OctueJSONDecoder`  
- [x] Remove unnecessary methods from `LabelSet`
- [x] Rename `add_labels` method and add `add` method to `LabelSet`

### Fixes
- [x] Handle timestamps from cloud with/without timezone information
- [x] Fix `OctueJSONDecoder`
- [x] Make it harder to add invalid labels to `LabelSet` 

### Dependencies
- [x] Use new version of `twined` that distinguishes tags from labels

### Testing
- [x] Use latest GCS emulator

## Quality Checklist
- [x] New features are fully tested (No matter how much Coverage Karma you have)
- [x] New features are documented

### CODE REVIEW

File: docs/source/cloud_storage_advanced_usage.rst [hunk 1]
```diff
@@ -26,14 +26,14 @@ to any of these methods.
         local_path=<path/to/file>,
         bucket_name=<bucket-name>,
         path_in_bucket=<path/to/file/in/bucket>,
-        metadata={"tags": ["blah", "glah", "jah"], "cleaned": True, "id": 3}
+        metadata={"id": 3, "labels": ["blah", "glah", "jah"], "cleaned": True, "colour": "blue"}
```
Comment (thclark) [inline docs/source/cloud_storage_advanced_usage.rst:29]: isn't the id a uuid?

File: octue/cloud/storage/client.py [hunk 2]
```diff
@@ -97,6 +100,40 @@ def upload_from_string(
...
 
+    def get_metadata(self, cloud_path=None, bucket_name=None, path_in_bucket=None, timeout=_DEFAULT_TIMEOUT):
+        """Get the metadata of the given file in the given bucket. Either (`bucket_name` and `path_in_bucket`) or
+        `cloud_path` must be provided.
+
+        :param str|None cloud_path:
+        :param str|None bucket_name:
+        :param str|None path_in_bucket:
+        :param float timeout:
+        :return dict:
```
Comment (thclark) [inline octue/cloud/storage/client.py:111]: docstring

File: octue/cloud/storage/client.py [hunk 3]
```diff
@@ -250,6 +255,19 @@ def _update_metadata(self, blob, metadata):
         :param dict metadata:
```
Comment (thclark) [inline octue/cloud/storage/client.py:255]: docstring

File: octue/resources/dataset.py [hunk 4]
```diff
@@ -223,18 +220,12 @@ def get_sequence_number(file):
...
-    def get_file_by_tag(self, tag_string):
-        """Gets a data file from a manifest by searching for files with the provided tag(s)
+    def get_file_by_label(self, label_string):
+        """Get a single datafile from a dataset by searching for files with the provided label(s).
 
         Gets exclusively one file; if no file or more than one file is found this results in an error.
 
-        :param tag_string: if this string appears as an exact match in the tags
-        :return: DataFile object
+        :param str label_string:
```
Comment (thclark) [inline octue/resources/dataset.py:228]: docstring

File: octue/resources/filter_containers.py [hunk 5]
```diff
@@ -1,33 +1,127 @@
...
+        for filter_name in filter_names:
+            filter_value = kwargs.pop(filter_name)
+            return self.filter(raise_error_if_filter_is_invalid, **{filter_name: filter_value}).filter(**kwargs)
+
+    def order_by(self, attribute_name, reverse=False):
+        """Order the `Filterable`s in the container by an attribute with the given name, returning them as a new
+        `FilterList` regardless of the type of filter container begun with.
+
+        :param str attribute_name:
+        :param bool reverse:
```
Comment (thclark) [inline octue/resources/filter_containers.py:42]: docstring

File: octue/resources/label.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            raise InvalidLabelException(
+                f"Invalid label '{cleaned_name}'. Labels must contain only characters 'a-z', 'A-Z', '0-9', and '-'. "
+                f"They must not start with '-'."
+            )
+
+        return cleaned_name
+
+
+class LabelSet(set):
+    """Class to handle a set of labels as a string."""
```
Comment (thclark) [inline octue/resources/label.py:46]: docstring needs param labels for __init__

File: octue/resources/label.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    def any_label_contains(self, value):
+        """ Return True if any of the labels contains value. """
+        return any(value in label for label in self)
+
+    def serialise(self, to_string=False, **kwargs):
+        """Serialise to a sorted list of label names.
+
+        :param bool to_string:
+        :return list|str:
+        """
```
Comment (thclark) [inline octue/resources/label.py:111]: docstrings with (brief) parameter description for the public methods

File: octue/resources/label.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+            return string
+
+        return json.loads(string)
+
+    @classmethod
+    def deserialise(cls, serialised_labelset):
+        """Deserialise from a sorted list of label names.
+
+        :param list serialised_labelset:
+        :return LabelSet:
```
Comment (thclark) [inline octue/resources/label.py:124]: docstring

File: octue/resources/tag.py [hunk 9]
```diff
@@ -1,195 +1,58 @@
...
+                    f"not start with '_'."
+                )
 
     def serialise(self, to_string=False, **kwargs):
-        """Serialise to a sorted list of tag names.
+        """Serialise a TagDict to a JSON dictionary or string.
 
         :param bool to_string:
-        :return list|str:
+        :return str|dict:
```
Comment (thclark) [inline octue/resources/tag.py:51]: docstring

**Review by thclark — CHANGES_REQUESTED** [review]
All the docstrings ;)

Code is great though!

**Comment by cortadocodes** [comment]
> All the docstrings ;)
> 
> Code is great though!

🙈  I should refer myself to the docstring doctor 👨‍⚕️

**Comment by thclark** [comment]
> > All the docstrings ;)
> > Code is great though!
> 
> 🙈 I should refer myself to the docstring doctor 👨‍⚕️

You should refer to yourself as the docstring vortex 🌪️ ;)

File: docs/source/cloud_storage_advanced_usage.rst (hunk 1 above)
Comment (cortadocodes) [reply docs/source/cloud_storage_advanced_usage.rst:29]: It is for `Identifiables` but this is just metadata for any given file (this isn't a method of an `Identifiable`)

File: octue/mixins/filterable.py [hunk 10]
```diff
@@ -32,7 +33,7 @@ def generate_complementary_filters(name, func):
 IN_RANGE_FILTER_ACTIONS = generate_complementary_filters("in_range", lambda item, value: value[0] <= item <= value[1])
 
 ICONTAINS_FILTER_ACTIONS = generate_complementary_filters(
-    "icontains", lambda item, value: value.lower() in item.lower()
+    "icontains", lambda item, value: value.casefold() in item.casefold()
```
[the remark this answers is not in the document]
Comment (cortadocodes) [reply octue/mixins/filterable.py:36]: 😂

**Pull request merged**